tls/datadog_checks/tls/tls.py

```diff
@@ -0,0 +1,348 @@
+
+        self._server = parsed_uri.hostname
+
+        # TODO: Support (implement) UDP
+        # https://chris-wood.github.io/2016/05/06/OpenSSL-DTLS.html
+        transport = self.instance.get('transport', 'tcp').lower()
+        if transport == 'udp':
+            # SOCK_DGRAM
+            self._sock_type = socket.SOCK_STREAM
+            # Default to 4433 (no standard port, but it's what OpenSSL uses)
+            self._port = int(self.instance.get('port', parsed_uri.port or 443))
+        else:
+            self._sock_type = socket.SOCK_STREAM
+            self._port = int(self.instance.get('port', parsed_uri.port or 443))
+
+        # https://en.wikipedia.org/wiki/Server_Name_Indication
+        self._server_hostname = self.instance.get('server_hostname', self._server)
+        self._validate_hostname = is_affirmative(self.instance.get('validate_hostname', True))
+
+        self._cert = self.instance.get('cert')
+        if self._cert:
+            self._cert = expanduser(self._cert)
+
+        self._private_key = self.instance.get('private_key')
+        if self._private_key:
+            self._private_key = expanduser(self._private_key)
+
+        self._cafile = None
+        self._capath = None
+        ca_cert = self.instance.get('ca_cert')
+        if ca_cert:
+            ca_cert = expanduser(ca_cert)
+            if isdir(ca_cert):
+                self._capath = ca_cert
+            else:
+                self._cafile = ca_cert
+
+        # Thresholds expressed in seconds take precedence over those expressed in days
+        self._seconds_warning = (
+            int(self.instance.get('seconds_warning', 0))
+            or days_to_seconds(float(self.instance.get('days_warning', 0)))
+            or self.DEFAULT_EXPIRE_SECONDS_WARNING
+        )
+        self._seconds_critical = (
+            int(self.instance.get('seconds_critical', 0))
+            or days_to_seconds(float(self.instance.get('days_critical', 0)))
+            or self.DEFAULT_EXPIRE_SECONDS_CRITICAL
+        )
+
+        # https://docs.python.org/3/library/ssl.html#ssl.SSLSocket.version
+        self._allowed_versions = get_protocol_versions(self.instance.get('allowed_versions', []))
+
+        # Global tags
+        self._tags = self.instance.get('tags', [])
+        if self._name:
+            self._tags.append('name:{}'.format(self._name))
+
+        # Decide the method of collection for this instance
+        if self._local_cert_path:
+            self.check = self.check_local
+            if self._validate_hostname and self._server_hostname:
+                self._tags.append('server_hostname:{}'.format(self._server_hostname))
+        else:
+            self.check = self.check_remote
+            self._tags.append('server_hostname:{}'.format(self._server_hostname))
+            self._tags.append('server:{}'.format(self._server))
+            self._tags.append('port:{}'.format(self._port))
+
+        # Cache lazily
```

Comment:
Let's comment and explain what these caches are for and what their structure look like.